fluids/datasaver.py

```diff
@@ -0,0 +1,78 @@
+from fluids.actions import SteeringAccAction
+from fluids.consts import OBS_NONE
+from fluids.utils import *
+import pickle
+import os
+
+class DataSaver():
+    """
+    Saves data in pickle file. After loading from pickle, data is in following format
+    dict[time]["obs"][OBS_NAME][key]
+    dict[time]["act"][ACT_NAME][key]
+    """
+
+    def __init__(self, fluid_sim, file, keys=None, obs=[OBS_NONE], act=[SteeringAccAction], batch_size=500, make_dir=True):
+        """
+            Save data from FLUIDS simulation.
+
+            fluid_sim (required): FluidSim data that is being saved
+            file (required): Filename with path to dump data to. Multiple files 
+                             with suffixes _1, _2 etc. might be created if data
+                             gets too large
+            keys: Keys of cars to gather observations from. Default is controlled car keys.
+            obs: List of observations to record. Default is [OBS_NONE]
+            act: List of actions to record. Default is [SteeringAccAction]
+            batch_size: Number of iterations to run before dumping data. Default is 500.
+            make_dir: Boolean. Flags if directory specified should be created or not.
+        """
+        self.fluid_sim = fluid_sim
+        self.file = file
+        self.obs = obs
+        self.act = act
+        self.batch_size = batch_size
+        self.keys = keys
+        if self.keys == None:
+            self.keys = self.fluid_sim.state.background_cars.keys()
+        if make_dir:
+            dir = os.path.dirname(self.file)
+            os.makedirs(dir, exist_ok=True)
+
+        self.curr_batch = 0
+        self.file_num = 0
+        self.curr_data = {}
+        self.separator = "_"
+
+    def dump(self):
+        file_name = "{}{}{}{}".format(self.file, self.separator, self.file_num, ".pickle")
+        fluids_print("Dumping batch in {}".format(file_name))
+
+
+        with open(file_name, 'wb') as handle:
+            pickle.dump(self.curr_data, handle, protocol=pickle.HIGHEST_PROTOCOL)
```

Comment:
I'm surprised this produces dumps as large as you say they are. `np.save` is a good alternative if pickling does not work out.

Reply:
np.save is giving me essentially identical performance, although I suspect there is not a lot of compression happening. This is because when I compress the 750mb file to a zip, I get a 6mb file back out. We could try something like this (https://stackoverflow.com/questions/22400652/compress-numpy-arrays-efficiently) directly?

Reply:
Hmm what are the parameters of the observation? Like type, resolution, etc. I want to calculate if 750 MB is reasonable for what we are collecting.

Compressing it is a good idea. I think our data is pretty compressible.

Reply:
Here's the breakdown for how much data we should be getting. I'm only considering the observation because it's way larger than the action space. The test I did was for the birds_eye_observation which consister of (500,500,3) array of uint8s. 
This means the total data should be 500*500*3*8bits*10cars * 100 timesteps which is exactly 750 megabytes, so numpy does not compression while saving.

Reply:
.npz is 6mb for 100 timesteps, pushing the new code now. Using numpy as a storage mechanism does make the data storage a bit more complicated because I have to define a custom data type for the numpy array, but I still think it's better than the pickle based approach because it seems cleaner to interface with.

Reply:
One downside is that it takes about 30 seconds to compress the file each time, so we should decide what batch size to dump data at to maximize speed. I can also test with other compression schemes sometime tomorrow, so it's probably best to hold off merging the pull request until then

Reply:
Sounds good, thanks for doing this analysis. 

I think we will probably usually be dumping the GridObservations anyways, not the BirdsEye, so maybe that will decrease the data rate significantly.